Issues/KP-324 cours links in pdf

amirhossein-haerian opened the pull request.

belanglos reviewed (commented):
This looks like it fixes the issue. Question: is there a scenario in which we _want_ to use the `courseLink` function?

belanglos commented:
Right, so what I'm thinking is, in those other places, do we not want the same behavior as here? In case somebody finds an `app.*` endpoint, we do not want the links to go to `app.`.

amirhossein-haerian commented:
> Right, so what I'm thinking is, in those other places, do we not want the same behavior as here? In case somebody finds an `app.*` endpoint, we do not want the links to go to `app.`.

That is completely true and it was a good point that you mentioned, I was thinking about that as well and I would say there is not only coursLinks that needs to be changed in that case; all of the links now use the origin as the base url and go to new directions based on the origin.

I can work on them and make the required changes for all of the links which will be a huge change and needs comprehensive testing but this is feasible completely. If you think it should be done with this task I can start working on that 🙌🚀

amirhossein-haerian commented:
> > Right, so what I'm thinking is, in those other places, do we not want the same behavior as here? In case somebody finds an `app.*` endpoint, we do not want the links to go to `app.`.
> 
> That is completely true and it was a good point that you mentioned, I was thinking about that as well and I would say there is not only coursLinks that needs to be changed in that case; all of the links now use the origin as the base url and go to new directions based on the origin.
> 
> I can work on them and make the required changes for all of the links which will be a huge change and needs comprehensive testing but this is feasible completely. If you think it should be done with this task I can start working on that 🙌🚀

I found another solution which can be found in the new changes that I have made in this PR, I realized  almost all of the links has been created using pageLink so I tried to change the domain there when it starts with app. we need to consider not to change the domain for admin links if we want to use this approach in other repositories as well.

axelbjo reviewed (approved):
LGTM!

amirhossein-haerian commented:
> Hm, was the initial problem that links in the PDF links to app.*? I'm not sure I really like the idea to rewrite all links for the web version as well. Lets have a sync about this after standup tomorrow.

The initial problem for this task was the problem with opening PDFs when the URL was starting with the app instead of www;
we fixed that by changing the API call URL to an app when the window.origin begins with the app.

Then we realized that links from pdf go to URLs beginning with the app if it is created from a URL starting with the app so first I fixed the issue for PDFs only but then we realized that the problem is for all the links in the web version, not only inside the pdf.

So the last changes that I made will fix the issue for the whole web and now by using the latest changes if you are using the services from the URLs beginging with the app and you use a link the link will send you to the actual URLs which begin with www.

We can discuss it further in the sync.

karlandindrakryggen commented on public/js/app/util/links.js:
```diff
@@ -40,7 +46,8 @@ function programmeWebLink(programmeCode, language) {
 
 function appendix1Link(programmeCode, term) {
-  return `/student/kurser/program/${programmeCode}/${term}/kurslista`
+  const baseUrl = possiblyChangeBaseUrl()
+  return `${baseUrl}/student/kurser/program/${programmeCode}/${term}/kurslista`
```
```suggestion
  return pageLink(`/student/kurser/program/${programmeCode}/${term}/kurslista`)
```
I think this is the only place you use `possiblyChangeBaseUrl` outside of `domain/links.js`. So test if it is possible to use `pageLink()` here as well.

amirhossein-haerian replied:
This was a great comment ⭐️ I made the required changes and tested it in my local environment; everything looks good.

karlandindrakryggen reviewed (changes requested):
Okay, lets go with this solution you already implemented as we discussed at sync.

Can you a look at if it is possible to keep `possiblyChangeBaseUrl()` as an internal function in `domain/links.js` (eg not export it at all) and instead only use `pageLink()` in other places? I think it's a good idea to keep a special solution like `possiblyChangeBaseUrl` as isolated as possible.
